### PR TITLE
Add nested tracing and endpoint filter tracing to apiserver 

### PR DESCRIPTION
**Pull request by jpbetz**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

Leverage https://github.com/kubernetes/utils/pull/139 to nest traces.

Also add tracing to the filter chain so that http requests are traced comprehensively. Includes adding the authentication and authorization checks as a steps in the HTTP level trace.

Example of the tracing output when an etcd list operation, authn and authz are slow:

```
Trace[1374095871]: "HTTP Request" method:GET,url:/api/v1/namespaces/kube-system/resourcequotas,user-agent:kube-apiserver/v1.19.0 (linux/amd64) kubernetes/482c37a,client:::1 (02-Jun-2020 10:38:00.615) (total time: 3201ms):
Trace[1374095871]: ---"Authenticate check done" 1000ms (10:38:00.615)
Trace[1374095871]: ---"Authorize check done" 1000ms (10:38:00.616)
Trace[1374095871]: ["List" url:/api/v1/namespaces/kube-system/resourcequotas,user-agent:kube-apiserver/v1.19.0 (linux/amd64) kubernetes/482c37a,client:::1 1200ms (10:38:00.616)
Trace[1374095871]:  ["List etcd3" key:/resourcequotas/kube-system,resourceVersion:,limit:0,continue: 1200ms (10:38:00.616)]]
Trace[1374095871]: [3.201532894s] [3.201532894s] END
```

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Trace output in apiserver logs is more organized and comprehensive. Traces are nested, and for all non-long running request endpoints, the entire filter chain is instrumented (e.g. authentication check is included).
```
/sig api-machinery
/priority important-longterm
/cc @SomtochiAma @lavalamp

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go [hunk 1]
```diff
@@ -33,21 +33,23 @@ import (
...
 	"k8s.io/apiserver/pkg/util/dryrun"
 	utilfeature "k8s.io/apiserver/pkg/util/feature"
-	utiltrace "k8s.io/utils/trace"
 )
 
 func createHandler(r rest.NamedCreater, scope *RequestScope, admit admission.Interface, includeName bool) http.HandlerFunc {
 	return func(w http.ResponseWriter, req *http.Request) {
 		// For performance tracking purposes.
-		trace := utiltrace.New("Create", utiltrace.Field{Key: "url", Value: req.URL.Path}, utiltrace.Field{Key: "user-agent", Value: &lazyTruncatedUserAgent{req}}, utiltrace.Field{Key: "client", Value: &lazyClientIP{req}})
+		trace := utiltrace.New("Create", utiltrace.Field{Key: "url", Value: req.URL.Path}, utiltrace.Field{Key: "user-agent", Value: &internal.LazyTruncatedUserAgent{req}}, utiltrace.Field{Key: "client", Value: &internal.LazyClientIP{req}})
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: `internal` 😍

**Comment by jpbetz** [comment]
rebased on master now that #87610 is merged

/hold cancel

**Comment by jpbetz** [comment]
pull-kubernetes-verify timed out

/retest

**Comment by jpbetz** [comment]
@lavalamp, @sttts Would either of you be willing to review?

**Comment by jpbetz** [comment]
/assign @logicalhan 

for review

/assign @ahg-g @sttts

for approval once reviewed

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/context.go [hunk 2]
```diff
@@ -94,3 +98,17 @@ func AuditEventFrom(ctx context.Context) *audit.Event {
...
+// TraceFrom returns the value of the trace key on the ctx
+func TraceFrom(ctx context.Context) (*utiltrace.Trace, bool) {
+	trace, ok := ctx.Value(traceKey).(*utiltrace.Trace)
+	return trace, ok
+}
+
+// WithTrace returns a new trace using the provided msg and fields, nested within any trace already in the
+// provided context. Also returns a context containing the new trace
+func WithTrace(ctx context.Context, msg string, fields ...utiltrace.Field) (context.Context, *utiltrace.Trace) {
+	parent, _ := TraceFrom(ctx)
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/endpoints/request/context.go:111]: Don't we need to check the second return value here? Looking above it seems like it is possibly nil.

Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/endpoints/request/context.go:111]: This is deliberate. The below `parent.Nest` can be called with a `nil` receiver. I'll add a comment explaining.

Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/endpoints/request/context.go]: Comment added.

**Comment by jpbetz** [comment]
Rebased and feedback applied.

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 3]
```diff
@@ -682,6 +682,7 @@ func DefaultBuildHandlerChain(apiHandler http.Handler, c *Config) http.Handler {
 	handler = genericapifilters.WithAuditAnnotations(handler, c.AuditBackend, c.AuditPolicyChecker)
 	handler = genericapifilters.WithWarningRecorder(handler)
 	handler = genericapifilters.WithCacheControl(handler)
+	handler = genericapifilters.WithTrace(handler, c.LongRunningFunc)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/config.go:685]: don't you have to be before WithRequestInfo? These filters are like shells added to the outside. I.e. requests hit the lower filters first and inner filters last.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/trace.go [hunk 4]
```diff
@@ -50,3 +51,9 @@ func WithTrace(handler http.Handler, longRunningCheck genericapirequest.LongRunn
 		handler.ServeHTTP(w, req)
 	})
 }
+
+func traceFilterStep(ctx context.Context, msg string, field ...utiltrace.Field) {
+	if trace, ok := genericapirequest.TraceFrom(ctx); ok {
+		trace.Step(msg, field...)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/trace.go:57]: we should have added nil receiver handling to `Step(...)`.

**Comment by sttts** [comment]
Looks pretty good. https://github.com/kubernetes/kubernetes/pull/88936/files#r448459806 is serious.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/trace.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+				isLongRunning = true
+			}
+		}
+
+		if !isLongRunning {
+			trace := utiltrace.New("HTTP "+req.Method,
+				utiltrace.Field{Key: "url", Value: req.URL.Path},
+				utiltrace.Field{Key: "user-agent", Value: &internal.LazyTruncatedUserAgent{req}},
+				utiltrace.Field{Key: "client", Value: &internal.LazyClientIP{req}})
+			defer trace.LogIfLong(750 * time.Millisecond)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/trace.go:49]: is this really the number we used before? Thinking especially about LIST requests which will never finish in that time. Didn't we define the time bound in the endpoint implementations before?

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/trace.go:49]: we could also have a much larger number here and let the endpoints shrinken it down for all the steps inside the endpoint implementation.

File: staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 3 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/server/config.go:685]: Yes, you're right, we want request info to be available so I need to move this up. Looks like the trace filter will still bracket all the other filters that might incur latency if I do this, which is the main reason I've tried to make one of the "outermost" filters initially.

Fixing.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/trace.go (hunk 5 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/trace.go:49]: Good question.

Did a grep for "LogIfLong" in endpoint and they're all 500ms. create, get, update, patch, delete, and [list](https://github.com/jpbetz/kubernetes/blob/4439b767f5b27a7931a493566233c6672203c2c5/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go#L277).

I agree 500ms seems low for list.

How about we pick a number that we feel is appropriate for list's and set the filter limit to be maybe 500ms over that number?

Maybe 1s for list and 1.5s for the trace filter?

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/trace.go (hunk 4 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/trace.go:57]: yeah, missed that

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/trace.go (hunk 5 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/trace.go:49]: Don't we still have the lower numbers in the endpoints? This is just another bound. If a nested trace exceeds the nested bound, we still get logging, don't we? So even if we put a 30s here, we are fine.

Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/trace.go:49]: Yeah, we do. I'm okay with starting with 30s, I guess. That way we don't risk accidentally flooding the logs with this change.

**Comment by jpbetz** [comment]
Thanks for the review @sttts. Feedback applied and commits squashed.

**Comment by jpbetz** [comment]
tests timed out

/retest

**Comment by jpbetz** [comment]
tests timed out

/retest

**Comment by jpbetz** [comment]
@sttts All feedback applied. This one is ready for another pass.

**Comment by jpbetz** [comment]
@deads2k or @lavalamp Any way either of you could have a quick look at this one? It's been reviewed by @logicalhan and @sttts already. Was hoping to get it in for 1.19 if possible.

File: pkg/scheduler/core/generic_scheduler.go [hunk 6]
```diff
@@ -19,6 +19,7 @@ package core
 import (
 	"context"
 	"fmt"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
```
Comment (deads2k) [inline pkg/scheduler/core/generic_scheduler.go:22]: nit: order.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go [hunk 7]
```diff
@@ -259,7 +260,7 @@ func (c *webhookConverter) Convert(in runtime.Object, toGV schema.GroupVersion)
 		return out, nil
 	}
 
-	trace := utiltrace.New("Call conversion webhook",
+	ctx, trace := genericapirequest.WithTrace(context.TODO(), "Call conversion webhook",
```
Comment (deads2k) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go:263]: well, this how stands out nicely.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go [hunk 8]
```diff
@@ -21,6 +21,7 @@ package mutating
 import (
 	"context"
 	"fmt"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:24]: stefan will be disappointed if I don't point these out. You've got a couple.  TBH, I don't actually care, but others would appreciate ordering these.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/trace.go [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		isLongRunning := false
+		if longRunningCheck != nil {
+			ri, ok := genericapirequest.RequestInfoFrom(req.Context())
+			if ok && longRunningCheck(req, ri) {
+				isLongRunning = true
+			}
+		}
+
+		if !isLongRunning {
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/trace.go:42]: invert and reduce nesting?

```go
if isLongRunning{
    handler.ServeHTTP(w, req)
    return
} 

// the code in this block.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/trace.go [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+			ctx, trace := genericapirequest.WithTrace(req.Context(), "HTTP "+req.Method,
+				utiltrace.Field{Key: "url", Value: req.URL.Path},
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/trace.go:44]: we don't need a field for the req.Method for query purposes.

Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/trace.go:44]: should we also add some of the bits from the request info like verb, resource, group, version, etc?

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go [hunk 11]
```diff
@@ -61,7 +63,7 @@ const (
 func PatchResource(r rest.Patcher, scope *RequestScope, admit admission.Interface, patchTypes []string) http.HandlerFunc {
 	return func(w http.ResponseWriter, req *http.Request) {
 		// For performance tracking purposes.
-		trace := utiltrace.New("Patch", utiltrace.Field{Key: "url", Value: req.URL.Path}, utiltrace.Field{Key: "user-agent", Value: &lazyTruncatedUserAgent{req}}, utiltrace.Field{Key: "client", Value: &lazyClientIP{req}})
+		req, trace := requestWithTrace(req, "Patch", utiltrace.Field{Key: "url", Value: req.URL.Path}, utiltrace.Field{Key: "user-agent", Value: &internal.LazyTruncatedUserAgent{req}}, utiltrace.Field{Key: "client", Value: &internal.LazyClientIP{req}})
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:66]: I'm trying to work this out.  Are there other places we mutate out inbound request's context in this way after the filter chain?

> To create a new request with a context, use NewRequestWithContext. To change the context of a request, such as an incoming request you want to modify before sending back out, use Request.Clone. Between those two uses, it's rare to need WithContext.

I had to look up doc on what this method was doing: https://golang.org/pkg/net/http/#Request.WithContext

Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:66]: also, aren't these fields duplicates from the main handler?

**Comment by deads2k** [comment]
some nits on order and preferences for formating.  A question about how "normal" a thing is and some questions about additional fields to add.

Nothing that I would hold it out of 1.19 for, but I would like the answers and nits in a followup bug fix please.

/lgtm
/approve
/milestone v1.19

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go (hunk 8 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:24]: Thanks. This PR predates we fixing my editor to do the right thing automatically :/ Fixed.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/trace.go (hunk 10 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/trace.go:44]: That would be much better. Adding.

Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/trace.go:44]: Going with:

```
utiltrace.Field{Key: "method", Value: req.Method},
utiltrace.Field{Key: "url", Value: req.URL.Path},
utiltrace.Field{Key: "verb", Value: ri.Verb},
utiltrace.Field{Key: "name", Value: ri.Name},
utiltrace.Field{Key: "resource", Value: ri.Resource},
utiltrace.Field{Key: "subresource", Value: ri.Subresource},
utiltrace.Field{Key: "namespace", Value: ri.Namespace},
utiltrace.Field{Key: "api-group", Value: ri.APIGroup},
utiltrace.Field{Key: "api-version", Value: ri.APIVersion},
utiltrace.Field{Key: "user-agent", Value: &internal.LazyTruncatedUserAgent{req}},
utiltrace.Field{Key: "client", Value: &internal.LazyClientIP{req}})
```

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go (hunk 11 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:66]: Good catch. Looks like I need to use `Request.Clone`.

Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:66]: They are. The catch is that they don't always both get included in the trace. The handler level trace output is only printed if the 30s threshold is exceeded, and the operation trace output (in this case "patch") is only printed if this operation handler exceeds 750ms. So when they both get output, there will be some redundant output, but when only one of them is printed, it is all needed.

**Comment by jpbetz** [comment]
> some nits on order and preferences for formating. A question about how "normal" a thing is and some questions about additional fields to add.
> 
> Nothing that I would hold it out of 1.19 for, but I would like the answers and nits in a followup bug fix please.

Thanks for the quick review! I've gone ahead and fixed the imports and request.WithContext calls with request.Clone and answered some of the questions. I added some more obvious field but there still might need to do a follow up on that part.

> 
> /lgtm
> /approve
> /milestone v1.19

**Comment by jpbetz** [comment]
/retest

Looks like verify didn't get far enough to output a build log

**Comment by lavalamp** [comment]
/approve

for the other directories.

**Comment by deads2k** [comment]
/kind bug
/priority important-soon

This is fixing the trace formatting to be in-line with expected output

**Pull request merged**